SRv6 TI-LFA (Topology Independent Loop-Free Alternate) support for IS-IS

vjardin opened the pull request:
## TI-LFA, SRLG

The implementation computes SRv6 segment stacks containing End SIDs for P-nodes and End.X SIDs for Q-node adjacencies, allowing traffic to be steered around failures through pre-computed backup paths. Multi-hop backup paths are supported when the backup route requires traversing intermediate nodes before reaching the post-convergence path.

The existing TI-LFA code keeps a clean architecture for maintaining both data planes with MPLS or SRv6. But I do focus on SRv6.

SRLG (Shared Risk Link Group) protection has been added to ensure backup paths avoid links that share the same failure risk as the protected link. This includes SRLG sub-TLV parsing, interface configuration commands, and integration with Flex-Algo constraints.

I keep adding weight-based selection for End.X SIDs and algorithm filtering to ensure compatibility with Flex-Algo deployments.

## Topotests and unit testing

It has been validated with the existing ISIS unit tests and a new grid topology stress test.

Four new topotests cover the main SRv6 TI-LFA scenarios: basic backup paths, SRLG protection, node protection, and multi-hop backup paths. All existing ISIS topotests continue to pass.

This serie includes a first commit to allow topotests to be able to run using a local setup of FRR which is convenient on compilation servers.

vjardin committed:
isisd: add SRv6 TI-LFA data structures

Add data structures to support Topology Independent Loop-Free Alternate
(TI-LFA) fast reroute using SRv6 instead of MPLS labels.

New structures added:
- isis_tilfa_srv6_sid_type: enum for End SID and End.X SID types
- isis_tilfa_srv6_sid: SRv6 SID entry containing the IPv6 SID value
  and source node system ID
- isis_srv6_seg_stack: segment stack holding up to SRV6_MAX_SEGS
  IPv6 segments for backup paths

The srv6_seg_stack field is added to isis_vertex_adj and isis_nexthop
structures to store SRv6 backup path information alongside the existing
MPLS label_stack field.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add remote SRv6 SID lookup functions for TI-LFA

Implement functions to look up SRv6 SIDs from remote nodes' LSPs,
which is required for building TI-LFA backup paths with SRv6.

New functions:
- isis_srv6_tilfa_find_pnode_end_sid(): Find End SID for a P-node by
  iterating through TLV 27 (SRv6 Locator) in the node's LSP. Accepts
  End, End with PSP, and uN (Next C-SID) behaviors.

- isis_srv6_tilfa_find_qnode_endx_sid(): Find End.X SID from source
  to neighbor by checking Sub-TLV 43 (SRv6 End.X SID) for P2P
  adjacencies and Sub-TLV 44 (SRv6 LAN End.X SID) for broadcast
  adjacencies in Extended IS Reachability TLV 22.

These functions are the SRv6 equivalents of the existing MPLS
tilfa_find_pnode_prefix_sid() and tilfa_find_qnode_adj_sid().

Also adds tilfa_enabled configuration flag to isis_srv6_db.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: implement SRv6 TI-LFA repair list building

Add core SRv6 TI-LFA implementation that builds repair paths using
SRv6 End and End.X SIDs instead of MPLS labels.

New functions:
- tilfa_compute_srv6_seg_stack(): Convert a repair list of SRv6 SIDs
  into a segment stack. Unlike MPLS which requires SRGB translation,
  SRv6 SIDs are direct IPv6 addresses.

- tilfa_srv6_repair_list_apply(): Apply the computed SRv6 segment
  stack to destination vertex adjacencies.

- tilfa_build_srv6_repair_list(): Recursive function that walks the
  post-convergence SPF tree to build the repair list. Uses End SIDs
  for P-nodes and End.X SIDs for the final hop to Q-nodes.

Modify isis_tilfa_check() to select between MPLS and SRv6 TI-LFA
based on configuration. SRv6 TI-LFA requires:
- SRv6 enabled with a locator configured
- tilfa_enabled flag set
- IPv6 address family

The MSD check uses max_seg_left_msd from the SRv6 configuration.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add SRv6 segment stack support to route handling

This commit extends the route and nexthop infrastructure to carry
SRv6 segment stacks for TI-LFA backup paths, similar to how MPLS
label stacks are currently handled.

The isis_nexthop structure now stores an optional srv6_seg_stack
pointer alongside the existing label_stack. When a nexthop is
duplicated via isis_nexthop_dup(), the segment stack is copied
using the new srv6_seg_stack_dup() helper function. The memory
is properly freed in isis_nexthop_delete() and isis_vertex_adj_free().

The adjinfo2nexthop() function signature is extended to accept a
segment stack parameter, allowing the SRv6 backup path information
to flow from the SPF vertex adjacency through to the installed
nexthop. Existing MPLS SR code passes NULL for this parameter as
it does not use SRv6 segment stacks.

The ISIS_SRV6_SEG_STACK memory type is used for all segment stack
allocations to enable proper memory tracking.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add CLI and YANG support for SRv6 TI-LFA

This commit adds the configuration interface for enabling SRv6 TI-LFA
fast reroute protection on IS-IS.

The YANG model extends the segment-routing-srv6 container with a
fast-reroute/ti-lfa/enable leaf. When set to true, the TI-LFA
computation will use SRv6 End and End.X SIDs instead of MPLS labels
for building backup paths.

The CLI command is available under the segment-routing srv6 node:

  router isis 1
   segment-routing srv6
    locator LOC1
    fast-reroute ti-lfa

The "no fast-reroute ti-lfa" form disables SRv6 TI-LFA protection.

The northbound callback sets the tilfa_enabled flag in the area's
srv6db configuration structure and triggers LSP regeneration to
ensure the SPF recomputation picks up the new setting.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add algorithm filtering to SRv6 TI-LFA SID lookups

Per RFC 9352, SRv6 SIDs are associated with specific algorithms.
For End SIDs, the algorithm is specified in the SRv6 Locator TLV
(section 6). For End.X SIDs, the algorithm field is part of the
sub-TLV itself (sections 8.1 and 8.2).

This commit modifies the TI-LFA SRv6 SID lookup functions to filter
by the SPF tree's algorithm, which is critical for Flexible Algorithm
(Flex-Algo) support. Without this filtering, TI-LFA could select SIDs
from a different algorithm, resulting in incorrect backup paths.

For End SID lookup, the function now iterates directly over locator
TLVs instead of using the generic callback, allowing access to the
locator's algorithm field for filtering.

For End.X SID lookup, the algorithm field from the sub-TLV is now
compared against the SPF algorithm before selecting a SID.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add weight-based selection for SRv6 End.X SID lookup

Per RFC 9352 sections 8.1 and 8.2, SRv6 End.X SID sub-TLVs include a
weight field for load balancing purposes. RFC 8402 specifies that
higher weight values indicate preferred SIDs.

Previously, the TI-LFA End.X SID lookup would return the first
matching SID found. This could result in suboptimal SID selection
when multiple End.X SIDs are advertised for the same adjacency with
different weights.

This change modifies the lookup to iterate through all matching SIDs
and select the one with the highest weight value. This applies to
both P2P End.X SIDs (Sub-TLV 43) and LAN End.X SIDs (Sub-TLV 44).

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add SRLG protection framework for TI-LFA

Add framework support for Shared Risk Link Group (SRLG) protection
in TI-LFA. This allows computation of backup paths that avoid links
sharing the same SRLG values as the protected resource.

This commit adds the LFA_SRLG_PROTECTION enum to the protection type
enumeration and extends the protected resource structure with an
array to store up to 16 SRLG values. The protection type string
functions are updated to display SRLG information, and a new
spf_adj_check_srlg_affected() function provides the framework for
checking if an adjacency shares any SRLG with the protected resource.

The TI-LFA computation function isis_tilfa_compute() and the P-space
computation in vertex_is_affected() are updated to handle SRLG
protection similarly to node protection, populating and clearing the
list of affected nodes.

This is a framework implementation that prepares the codebase for
full SRLG protection support. The actual SRLG checking functionality
requires SRLG TLV (RFC 5307 sub-TLV 16) parsing which is not yet
implemented in FRR's IS-IS. Once SRLG TLV support is added to
isis_tlvs.c, the spf_adj_check_srlg_affected() function can be
completed to perform actual SRLG value comparisons.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add SRLG sub-TLV parsing for TI-LFA protection

Implement parsing and serialization of SRLG (Shared Risk Link Group)
sub-TLV as defined in RFC 5307 section 1.3. The SRLG sub-TLV uses type
16 which is also used by ASLA (RFC 9351), so a heuristic detection
function distinguishes between the two formats based on their structure.

The SRLG sub-TLV contains a list of 32-bit SRLG values that identify
links sharing common failure risks. This data is stored in the
isis_ext_subtlvs structure and used by the SRLG protection framework
in isis_lfa.c to compute SRLG-disjoint backup paths for TI-LFA.

The spf_adj_check_srlg_affected() function now uses the parsed SRLG
data to check whether an adjacency shares any SRLG value with the
protected resource, enabling proper SRLG-aware path computation.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add show isis srlg command

Add a dedicated command to display SRLG (Shared Risk Link Group)
information learned from the IS-IS LSP database. The command iterates
through Extended IS Reachability TLVs and displays SRLG values
associated with each neighbor adjacency.

The command syntax is:
  show isis [vrf <NAME|all>] srlg [level-1|level-2]

Example output:
  Area CORE:
    Level-2 SRLG information:
      0000.0000.0001 -> 0000.0000.0002: 100 200

This helps operators verify SRLG configuration for TI-LFA SRLG-disjoint
backup path computation.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add isis srlg interface configuration command

Add CLI command to configure SRLG values on IS-IS interfaces:

  interface eth0
    isis srlg <0-16777215>

The SRLG values are stored in circuit->ext and advertised in
Extended IS Reachability TLV (sub-TLV 16). This allows TI-LFA
to compute SRLG-disjoint backup paths.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add ti-lfa srlg-protection interface command

Add srlg-protection option to the TI-LFA interface command:

  isis fast-reroute ti-lfa [level-1|level-2] [srlg-protection]

When enabled, TI-LFA computes backup paths that are SRLG-disjoint
from the protected interface. The SRLG values are taken from the
circuit configuration (isis srlg command).

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: add SRLG constraint support to Flex-Algo

Implement the Exclude SRLG constraint for Flexible Algorithm as defined
in RFC 9350 Section 6.5. This allows network operators to exclude links
with specific SRLG values from Flex-Algo path computation, enabling
SRLG-disjoint routing topologies.

The implementation adds storage for up to 16 SRLG values in the flex_algo
structure and introduces a new CLI command "srlg exclude-any <value>" under
the flex-algo configuration node. The FAD (Flex-Algo Definition) TLV
parsing and encoding have been updated to handle the ESRLG sub-sub-TLV,
allowing SRLG constraints to be advertised and received from remote routers.

During SPF computation, links whose SRLG values match any of the excluded
SRLGs are dropped from the topology, ensuring paths avoid those shared
risk groups. The isis_flex_algo_supported() function has been updated to
accept definitions containing SRLG constraints, marking this feature as
fully supported.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: display SRv6 segment stack in show isis route

Extend the show isis route command to display SRv6 segment stacks for
backup nexthops using TI-LFA with SRv6. When a backup path has been
computed using SRv6 TI-LFA, the IPv6 SIDs in the segment stack are now
shown in the Label(s) column, separated by slashes, similar to how MPLS
label stacks are displayed.

This enhancement provides visibility into the SRv6 repair paths that
have been computed for fast reroute protection, allowing operators to
verify that TI-LFA SRv6 backup paths are correctly installed with the
expected End and End.X SIDs.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: extend unit tests for SRv6 and SRLG support

The IS-IS unit test infrastructure has been extended to support testing
of SRv6 TI-LFA and SRLG protection features. The test_common module now
includes the ability to generate SRv6 TLVs including SRv6 Locator with
End SID and SRv6 End.X SID sub-TLVs for adjacencies. Additionally, SRLG
sub-TLV generation has been added to Extended IS Reachability TLVs to
allow testing of SRLG-aware path computation.

Two new test topologies have been added. Topology 15 is a three-node
triangle where each node advertises an SRv6 locator with associated
End SID, enabling verification of SRv6 TI-LFA backup path computation.
Topology 16 is also a three-node triangle but with distinct SRLG values
assigned to each link, allowing verification that TI-LFA correctly
avoids links sharing the same SRLG as the protected resource.

The test data structures have been extended with isis_test_srv6 for
node-level SRv6 configuration and srlg arrays in isis_test_adj for
per-adjacency SRLG values. The F_ISIS_TEST_NODE_SRV6 flag enables SRv6
TLV generation for nodes that require it.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: fix TI-LFA backup path when nexthop lacks SR capability

When a router disables Segment Routing, TI-LFA backup paths through
that router were failing instead of using Implicit Null label.

Changes in tilfa_repair_list_apply():
- Check if backup nexthop has SR capability (SRGB)
- If nexthop lacks SR, use Implicit Null instead of computing label stack
- Allows backup path to work via IP forwarding from non-SR router

Changes in tilfa_build_repair_list():
- When Adj-SID lookup fails for P-node without SR, skip to P-node
  handling instead of returning error
- When P-node lacks Prefix-SID due to no SR capability, clear repair
  list and install backup with Implicit Null
- Prevents TI-LFA computation failure when P-node has no SR

Test expectation updates:
- rt2/step4: Remove 16060/16061 entries (routes via rt4 when rt4 has no SR)
- rt6/step4: Remove 16020/16021 entries (routes via rt4 when rt4 has no SR)
- Aligns with test comment: "All routers should uninstall all SR labels
  for destinations whose nexthop is rt4"

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add topotest for IS-IS SRv6 TI-LFA

This topotest validates the TI-LFA fast reroute functionality when using
SRv6 as the data plane instead of MPLS. The test topology mirrors the
existing isis_tilfa_topo1 test with a 6-router network where rt1, rt2,
and rt3 are connected via a shared LAN segment, rt2 connects to rt4 via
dual links, rt3 connects to rt5 via dual links, and rt4, rt5, rt6 form
a triangle at the bottom of the topology.

Each router is configured with an SRv6 locator using the uSID format
with PSP flavor. The IS-IS configuration enables TI-LFA protection on
the LAN interfaces and activates the SRv6 TI-LFA feature under the
segment-routing srv6 configuration node.

The test executes three steps. Step 1 verifies initial network
convergence by checking that all IS-IS adjacencies are established and
that IPv6 routes are installed with TI-LFA backup nexthops. Step 2
simulates a link failure by shutting down rt2's eth-sw1 interface and
verifies that adjacencies reconverge and routes switch to backup paths.
Step 3 restores the link and confirms that adjacencies and routes
return to their original state with TI-LFA protection re-established.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add topotest for IS-IS SRv6 TI-LFA with SRLG protection

Add a new topotest to verify SRLG (Shared Risk Link Group) protection
with SRv6 TI-LFA. This test extends the basic SRv6 TI-LFA test with
SRLG configuration:

- RT2 and RT4 interconnecting links (eth-rt4-1/2, eth-rt2-1/2) are
  assigned SRLG 100
- SRLG protection is enabled on these interfaces with
  "isis fast-reroute ti-lfa level-1 srlg-protection"
- Step 4 verifies SRLG values are advertised in IS-IS LSPs via
  "show isis srlg json"

With SRLG protection enabled, backup paths should avoid links sharing
the same SRLG, routing through RT3-RT5 instead of the alternate direct
link between RT2 and RT4.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add topotest for IS-IS SRv6 TI-LFA with node protection

Add a new topotest to verify node protection with SRv6 TI-LFA. This test
extends the basic SRv6 TI-LFA test with node protection configuration:

- RT1's eth-sw1 has node-protection enabled (protects against RT2 failure)
- RT3's eth-sw1 has node-protection enabled (protects against RT2 failure)
- RT4's eth-rt2-1 and eth-rt2-2 have node-protection enabled
- RT5's eth-rt3-1 and eth-rt3-2 have node-protection enabled

With node protection, backup paths protect not just against link failures
but also against node failures. For example, if RT2 fails entirely, RT1's
traffic to RT4 uses a backup path through RT3-RT5-RT4 instead of just
protecting against the RT1-RT2 link failure.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add topotest for IS-IS SRv6 TI-LFA multi-hop backup paths

Add a new topotest to verify multi-hop TI-LFA backup paths with SRv6
segment stacks. This test modifies the basic SRv6 TI-LFA topology by
removing the direct RT4-RT5 link, forcing backup paths to traverse
multiple hops through RT6.

Multi-hop TI-LFA Configuration:
- RT4-RT5 direct link is intentionally removed
- Backup paths from RT4 to RT5 must go through RT6
- This tests SRv6 segment stack computation with End SID (P-node)
  and End.X SID (Q-node adjacency)

Step 4 verifies IS-IS route output on RT4 and RT5 which require
multi-hop backup paths due to the missing direct link.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add unit test for multi-hop SRv6 TI-LFA (topology 17)

Add topology 17 to the IS-IS SPF unit test framework to verify
multi-hop TI-LFA backup path computation with SRv6.

Topology 17:

       RT1
       /   \
     RT2---RT3
      |     |
     RT4   RT5   (no direct link)
       \   /
        RT6

The topology has no direct link between RT4 and RT5, which forces
the TI-LFA algorithm to compute multi-hop backup paths through
intermediate nodes.

The unit test verifies:
- P-space computation: nodes reachable without using the protected link
- Q-space computation: nodes that can reach the destination without
  using the protected link
- Post-convergence SPF tree construction
- Backup path selection through the P-node/Q-node intersection

Test scenarios:
- From RT4, protect link to RT2: backup path via RT6->RT5->RT3
- From RT4, protect link to RT6: backup path via RT2->RT3->RT5
- From RT5, protect link to RT3: backup path via RT6->RT4->RT2
- From RT5, protect link to RT6: backup path via RT3->RT2->RT4

All nodes have SRv6 enabled with locators fc00:0:X::/48.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add grid topology generator for IS-IS SPF stress testing

Add a configurable grid topology generator for large-scale SPF and
TI-LFA stress testing in a single process (no network namespaces).

Grid topology layout (example 5x5):

  rt1 -- rt2 -- rt3 -- rt4 -- rt5
   |      |      |      |      |
  rt6 -- rt7 -- rt8 -- rt9 -- rt10
   |      |      |      |      |
  ...    ...    ...    ...    ...
   |      |      |      |      |
  rt21-- rt22-- rt23-- rt24-- rt25

Features:
- Runtime configurable dimensions (rows x cols, up to 256 nodes)
- Each node connects to up to 4 neighbors (up/down/left/right)
- Optional SRv6 support with unique locators per node
- Measures and reports SPF/TI-LFA computation time

New command syntax:
  test isis grid rows <1-256> cols <1-256> root HOSTNAME
       <spf|ti-lfa system-id WORD> [ipv4-only|ipv6-only] [srv6]

Example usage:
  test isis grid rows 10 cols 20 root rt1 spf ipv6-only srv6
  test isis grid rows 16 cols 16 root rt128 ti-lfa system-id rt127 srv6

Performance results (IPv6 SPF from corner node):
  - 5x5   (25 nodes):  ~2ms
  - 10x10 (100 nodes): ~10ms
  - 10x20 (200 nodes): ~15ms
  - 16x16 (256 nodes): ~12ms

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add SRv6 kernel support check for IS-IS SRv6 topotests

Add helper functions to check and enable SRv6 (seg6) kernel support
in topotests:

- check_kernel_seg6_support(): Check if kernel has seg6 sysctl available
  and whether it's currently enabled
- enable_srv6_on_router(): Enable seg6 on router interfaces (all, default, lo)

Update all IS-IS SRv6 topotests to use these helpers in setup_module():
- Skip test with clear message if kernel doesn't support SRv6
- Automatically enable seg6 on all routers before starting daemons

This fixes test failures on systems where net.ipv6.conf.all.seg6_enabled
defaults to 0.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: rename TI-LFA MPLS functions with mpls prefix

Rename TI-LFA MPLS-specific functions to include 'mpls' in their names
for consistency with the upcoming SRv6 TI-LFA implementation. This
establishes a clear naming convention where:
- MPLS functions use tilfa_mpls_* prefix
- SRv6 functions will use tilfa_srv6_* prefix

Signed-off-by: Vincent Jardin <vjardin@free.fr>

pguibert6WIND commented on isisd/isis_srv6.c:
```diff
@@ -805,8 +803,7 @@ isis_srv6_tilfa_find_pnode_end_sid(struct isis_spftree *spftree,
-	isis_lsp_iterate_srv6_locator(lsp, spftree->mtid,
-				      tilfa_srv6_find_end_sid_cb, &args);
+	isis_lsp_iterate_srv6_locator(lsp, spftree->mtid, tilfa_srv6_find_end_sid_cb, &args);
 
```
no need to do this change if nothing changed.
perhaps 'git clang-format HEAD~1' should have been done in the commit this change has been done.
it seems it is the case for many lines of this commit in isis_srv6.c

pguibert6WIND commented on isisd/isis_srv6.c:
```diff
@@ -736,6 +738,180 @@ void isis_srv6_area_term(struct isis_area *area)
+			args->found = true;
+			return LSP_ITER_STOP;
+		default:
+			/* Skip other behaviors (e.g., END.DT*, END.DX*) */
```
anyway, non matching defines should be added in the switch statement.

pguibert6WIND commented on tests/topotests/lib/topogen.py:
```diff
@@ -1038,8 +1046,9 @@ def vtysh_cmd(self, command, isjson=False, daemon=None):
-        vtysh_command = "vtysh {} -c {} 2>/dev/null".format(
-            dparam, shlex.quote(command)
+        vtysh_bin = os.path.join(self.net.daemondir, "vtysh")
+        vtysh_command = "{} {} -c {} 2>/dev/null".format(
```
with default settings, vtysh is not located on same folder as usr/lib/frr. please review this part of code.

vjardin replied:
thanks, I'll extract it this specific into a specific pull request. It is not related to this topic.

vjardin replied:
dropped, later pull request.

pguibert6WIND reviewed (changes requested):
partial review specific to the code.
I will need more time to have a look in details how the backup srv6 paths are used.

vjardin commented on tests/topotests/pytest.ini:
```diff
@@ -81,7 +81,7 @@ markers =
 # Default daemons binaries path.
-#frrdir = /usr/lib/frr
+frrdir = /mnt/256G/vjardin/dev/bin/sbin
 
```
right, not related, I'll drop this commit. I'll extract this commit and propose a specific release pull request.

vjardin committed:
tests: add unit test for SRv6 TI-LFA algorithm filtering

Add topology 18 to test that TI-LFA correctly filters SRv6 locators
by algorithm. Each node advertises two locators:
- fc00:0:X::/48 with algorithm 0 (SPF)
- fc00:128:X::/48 with algorithm 128 (Flex-Algo)

The test verifies that only algorithm 0 locators are used for
TI-LFA backup path computation, confirming the filtering logic in
tilfa_find_remote_srv6_end_sid() works correctly.

Note: This is a minimum test for the time being. A full topotest
cannot be added yet because the SRv6 locator algorithm is currently
hardcoded to 0 (SPF) in isis_lsp.c. When CLI support for configuring
SRv6 locators with different algorithms (Flex-Algo) is implemented,
a comprehensive topotest should be added.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin commented on isisd/isis_srv6.c:
```diff
@@ -736,6 +738,180 @@ void isis_srv6_area_term(struct isis_area *area)
+		switch (end_sid->behavior) {
+		case SRV6_ENDPOINT_BEHAVIOR_END:
+		case SRV6_ENDPOINT_BEHAVIOR_END_PSP:
+		case SRV6_ENDPOINT_BEHAVIOR_END_PSP_USP:
```
fixed.

vjardin commented on isisd/isis_srv6.c:
```diff
@@ -855,17 +863,22 @@ static int tilfa_srv6_find_endx_sid_cb(const uint8_t *id, uint32_t metric, bool
 			 * identifies the specific neighbor.
-			 * Skip backup SIDs.
+			 * Skip backup SIDs (B-flag).
 			 */
```
ok

RodrigoMNardi commented:
ci:rerun

pguibert6WIND commented on isisd/isis_spf.c:
```diff
@@ -2078,47 +1904,38 @@ static void isis_run_spf_cb(struct event *event)
-		for (ALL_LIST_ELEMENTS_RO(area->flex_algos->flex_algos, node,
-					  fa)) {
+		for (ALL_LIST_ELEMENTS_RO(area->flex_algos->flex_algos, node, fa)) {
 			data = fa->data;
```
ditto: warning of compilation with node unused

pguibert6WIND reviewed (changes requested):
still some compilation fixes to report on https://github.com/FRRouting/frr/pull/20351

pguibert6WIND committed:
isisd, lib: handle srv6 endx backup sids allocation

> rt4# show segment-routing srv6 sid
>  SID           Behavior    Context                      Daemon/Instance    Locator    AllocationType
>  ------------------------  ---------------------------  -----------------  ---------  ----------------
>  fc00:0:4::    uN          -                            isis(0)            loc1       dynamic
>  fc00:0:4:1::  uA          Interface 'eth-rt2-1'        isis(0)            loc1       dynamic
>  fc00:0:4:2::  uA          Interface 'eth-rt2-2'        isis(0)            loc1       dynamic
>  fc00:0:4:3::  uA          Interface 'eth-rt5'          isis(0)            loc1       dynamic
>  fc00:0:4:4::  uA          Interface 'eth-rt6'          isis(0)            loc1       dynamic
>  fc00:0:4:5::  uA          Interface 'eth-rt5' (b)      isis(0)            loc1       dynamic
>  fc00:0:4:6::  uA          Interface 'eth-rt2-1' (b)    isis(0)            loc1       dynamic
>  fc00:0:4:7::  uA          Interface 'eth-rt6' (b)      isis(0)            loc1       dynamic

Running isis_srv6_tilfa_topo1 triggers an asan error.

> 2026/01/09 10:11:03.666243 ISIS: lib/memory.c:79: mt_count_free(): assertion (mt->n_alloc) failed
> ISIS: Received signal 6 at 1767949863 (si_addr 0x7200001429, PC 0x75fff029eb2c); aborting...
> ISIS: /lib/x86_64-linux-gnu/libasan.so.8(+0x831e0) [0x75fff0c831e0]
> ISIS: /lib/libfrr.so.0(zlog_backtrace_sigsafe+0x1ac) [0x75fff08275ec]
> ISIS: /lib/libfrr.so.0(zlog_signal+0x2ad) [0x75fff0827bed]
> ISIS: /lib/libfrr.so.0(+0x2c814a) [0x75fff08c814a]
> ISIS: /lib/x86_64-linux-gnu/libc.so.6(+0x45330) [0x75fff0245330]
> ISIS: /lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x11c) [0x75fff029eb2c]
> ISIS: /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x1e) [0x75fff024527e]
> ISIS: /lib/x86_64-linux-gnu/libc.so.6(abort+0xdf) [0x75fff02288ff]
> ISIS: /lib/libfrr.so.0(_zlog_assert_failed+0x20e) [0x75fff094d51e]
> ISIS: /lib/libfrr.so.0(qfree+0x85) [0x75fff0833875]
> ISIS: /usr/lib/frr/isisd(isis_nexthop_delete+0x73) [0x571d961fb6f3]
> ISIS: /lib/libfrr.so.0(list_delete_all_node+0xb9) [0x75fff08110f9]
> ISIS: /lib/libfrr.so.0(list_delete+0x2d) [0x75fff081122d]
> ISIS: /usr/lib/frr/isisd(+0x127a0b) [0x571d961fba0b]
> ISIS: /lib/libfrr.so.0(route_table_finish+0x3cd) [0x75fff08e4a1d]
> ISIS: /usr/lib/frr/isisd(isis_spf_invalidate_routes+0x8d) [0x571d9621410d]
> ISIS: /usr/lib/frr/isisd(isis_area_invalidate_routes+0x84) [0x571d962730c4]
> ISIS: /usr/lib/frr/isisd(+0x13f860) [0x571d96213860]
> ISIS: /lib/libfrr.so.0(event_call+0x351) [0x75fff08fc1d1]
> ISIS: /lib/libfrr.so.0(frr_run+0x248) [0x75fff080ec48]
> ISIS: /usr/lib/frr/isisd(main+0x26e) [0x571d961aa80e]
> ISIS: /lib/x86_64-linux-gnu/libc.so.6(+0x2a1ca) [0x75fff022a1ca]
> ISIS: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x8b) [0x75fff022a28b]
> ISIS: /usr/lib/frr/isisd(_start+0x25) [0x571d961ac4b5]
> ISIS: in thread isis_run_spf_cb scheduled from isisd/isis_spf.c:2041 _isis_spf_schedule()

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

vjardin committed:
isisd: fix missing neighbor TLVs when MT is enabled with empty adjacency MT set

When Multi-Topology (MT) is enabled for an IS-IS area but the neighbor's
MT set from adjacency is empty (mt_count == 0), the tlvs_add_mt_set()
function would skip adding any neighbor TLVs. This breaks IS-IS routing
because LSPs are generated without Extended Reachability TLVs (TLV 22)
or MT Reachability TLVs (TLV 222).

The issue occurs because:
1. area_is_mt(area) returns true when MT is configured
2. The for loop iterating over mt_set never executes when mt_count is 0
3. No neighbor TLVs are added, resulting in incomplete LSPs

This can happen during initial convergence, adjacency flaps, or DIS
changes on broadcast networks when LSP regeneration occurs before
hello processing has populated the adjacency MT information.

Fix by falling back to the area's locally configured MT settings when
the adjacency MT set is empty. This ensures neighbor TLVs are always
generated for all enabled topologies.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: fix SRv6 TI-LFA backup path and add PSP behavior support

Fix SRv6 TI-LFA backup path computation to request End.X SID via SRv6
instead of SR-MPLS path. Add backup flag to srv6_sid_ctx for TI-LFA.

Add support for PSP (Penultimate Segment Pop) endpoint behaviors:
- SRV6_ENDPOINT_BEHAVIOR_END_PSP
- SRV6_ENDPOINT_BEHAVIOR_END_NEXT_CSID_PSP

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: update SRv6 TI-LFA node protection and SRLG test reference files

Simplify show_ipv6_route.ref files to check only the local SRv6
locator route. This verifies basic SRv6 functionality while avoiding
dependencies on remote route installation timing.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: fix memory issues in LFA backup route handling

Fix two related memory management issues:

1/ Double-free in adjinfo2nexthop(): The function was storing direct
pointers to label_stack and srv6_seg_stack from the vertex adjacency
list without copying. When the adjacency list was deleted after backup
route creation, these pointers became dangling. Fix by using
label_stack_dup() and srv6_seg_stack_dup() to create independent copies.

2/ Memory leak in isis_rlfa_activate(): When RLFA is reactivated for
the same vertex (e.g., receiving updated LDP responses), a new
label_stack was allocated without freeing the existing one. Fix by
freeing the old label_stack before allocating a new one.

Fixes: c951ee6eeeac ("isisd: add support for Topology Independent LFA (TI-LFA)")
Signed-off-by: Vincent Jardin <vjardin@free.fr>

cscarpitta commented on isisd/isis_zebra.c:
```diff
@@ -1522,6 +1523,9 @@ static int isis_zebra_srv6_sid_notify(ZAPI_CALLBACK_ARGS)
 				listnode_add(area->srv6db.srv6_sids, sid);
 
+				/* Regenerate the LSP to include End SID */
+				lsp_regenerate_schedule(area, area->is_type, 0);
```
This is redundant. We already call `lsp_regenerate_schedule` at line 1574

cscarpitta reviewed (changes requested):
A few initial comments.
 
- Please add documentation
- For new topotests, use `frr.conf` unified config

vjardin committed:
tests: add topotest for IS-IS SRv6 TI-LFA backup route verification

Topotest to validate that SRv6 segment stacks are correctly
displayed in the IS-IS backup routing table using the command
"show isis route backup json".

The test uses a diamond topology with four routers running IS-IS
level-1 with SRv6 and TI-LFA enabled:

            10.0.1.0/24         10.0.3.0/24
       .----------eth-rt2-----------.
       |          [RT2]             |
    [RT1]                         [RT4]
       |          [RT3]             |
       '----------eth-rt3-----------'
            10.0.2.0/24         10.0.4.0/24

Each router has a unique SRv6 locator (fc00:0:N::/48 where N is the
router number) with uSID behavior and PSP flavor enabled.

The test verifies backup path computation in three scenarios. First,
it checks that RT1 has computed backup paths through RT3 to reach
destinations behind RT2 when the RT1-RT2 link would fail. Second,
it verifies that after shutting down the RT1-RT2 interface, traffic
properly converges to the backup path through RT3 and RT4. Third,
it confirms that restoring the RT1-RT2 interface returns the network
to its original state with backup paths recomputed.

The reference JSON files validate adjacency formation, IPv6 route
installation with SRv6 nexthops, and backup route entries containing
the expected SRv6 segment stacks for loop-free alternate paths.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd: convert show_isis_srlg from DEFUN to DEFPY

Convert the show_isis_srlg command to use DEFPY for cleaner argument
parsing with automatic variable binding.

Since isis_te.c is compiled for both isisd and fabricd, but this command
is isisd-only, the clippy include requires a #ifndef FABRICD guard which
is an unusual pattern.

TBD: consider whether moving this command to isis_cli.c would be cleaner.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

pguibert6WIND committed:
isisd, lib: add seg6local backup contexts

> rt4# show segment-routing srv6 sid
>  SID           Behavior    Context                      Daemon/Instance    Locator    AllocationType
>  ------------------------  ---------------------------  -----------------  ---------  ----------------
> [..]
>  fc00:0:4:5::  uA          Interface 'eth-rt5' (b)      isis(0)            loc1       dynamic
>  fc00:0:4:6::  uA          Interface 'eth-rt2-1' (b)    isis(0)            loc1       dynamic
>  fc00:0:4:7::  uA          Interface 'eth-rt6' (b)      isis(0)            loc1       dynamic
>
> rt4# show ipv6 route
> [..]
> I>* fc00:0:4:5::/64 [115/0] is directly connected, eth-rt5, seg6local uA nh6 fe80::b026:36ff:fe26:cfe8, eth-rt5, weight 1, 00:00:03
>   b                           is directly connected, eth-rt6, seg6local uA nh6 fe80::b068:abff:fe1f:ae64, eth-rt6, weight 1
> I>* fc00:0:4:6::/64 [115/0] is directly connected, eth-rt2-2, seg6local uA nh6 fe80::f497:c1ff:fedb:6c7, eth-rt2-2, weight 1, 00:00:03
>   b                           is directly connected, eth-rt5, seg6local uA nh6 fe80::b026:36ff:fe26:cfe8, eth-rt5, weight 1
> I>* fc00:0:4:7::/64 [115/0] is directly connected, eth-rt6, seg6local uA nh6 fe80::b068:abff:fe1f:ae64, eth-rt6, weight 1, 00:00:03
>   b                           is directly connected, eth-rt5, seg6local uA nh6 fe80::b026:36ff:fe26:cfe8, eth-rt5, weight 1

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

vjardin committed:
tests: add path verification tests to SRv6 TI-LFA topotests

Add kernel route nexthop verification to the four SRv6 TI-LFA test
suites (topo1, srlg, node_protection, multihop) to ensure routes are
correctly installed and updated during link failure scenarios.

Each test file includes helper functions to query kernel routes
via iproute2 and extract nexthop information. The test sequence
captures baseline route nexthops before the link failure, verifies
that routes remain available after the failure, and confirms that
the original paths are restored after link recovery.

The traffic connectivity test in topo1 includes retry logic to
handle timing differences between FRR RIB convergence and kernel
route installation, ensuring reliable test execution.

This extends the existing control plane verification to also validate
the data plane programming, confirming SRv6 TI-LFA backup paths work
correctly end-to-end.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
doc: add SRv6 TI-LFA documentation to isisd user guide

SRv6 TI-LFA Design Overview
---------------------------

TI-LFA pre-computes backup paths using Segment Routing before failures
occur, enabling sub-50ms convergence when a link or node fails. The
implementation follows RFC 8402 (TI-LFA) with SRv6 dataplane per RFC 9352.

Unlike MPLS TI-LFA which uses label stacks requiring SRGB translation,
SRv6 TI-LFA uses IPv6 addresses directly as segment identifiers:

  - End SID: identifies a node (analogous to MPLS Prefix-SID)
  - End.X SID: identifies an adjacency (analogous to MPLS Adj-SID)

A key architectural difference is that SRv6 End.X SIDs are globally
routable. When traffic reaches the Q-node (adjacent to the destination
on the post-convergence path), normal IPv6 routing delivers the packet.
This eliminates the need to push a destination SID, resulting in shorter
segment stacks compared to MPLS TI-LFA.

Protection Types
----------------

Three protection levels are supported:

  - Link protection: backup avoids the failed link only
  - Node protection: backup avoids both the link and next-hop node
  - SRLG protection: backup avoids all links sharing SRLG values with
    the protected link (per RFC 5307)

SRLG (Shared Risk Link Group) protection addresses correlated failures
where links sharing physical resources (same conduit, bridge, etc.) fail
together. Each interface is assigned SRLG values (32-bit integers) which
are advertised in IS-IS LSPs via the SRLG sub-TLV. The TI-LFA algorithm
excludes any link whose SRLG set intersects with the protected link.

Configuration Hierarchy
-----------------------

SRv6 TI-LFA requires configuration at two levels:

  1. Global: under "segment-routing srv6" in router isis context
     - "fast-reroute ti-lfa" enables computation
     - "node-msd" sub-mode configures MSD advertisements

  2. Per-interface: "isis fast-reroute ti-lfa" with protection options
     - node-protection, link-fallback, srlg-protection

MSD (Maximum SID Depth) values advertise the router's segment processing
capabilities, ensuring backup paths don't exceed hardware limits.

Documentation includes the new "show isis srlg" command for displaying
SRLG information from the link-state database, and references existing
show commands for backup route and FRR summary display.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
zebra: add backup flag to SRv6 SID JSON output

Add missing json backup to xhe show segment-routing srv6 sid command
that displays backup End.X SIDs with a "(b)" suffix in the text output
to indicate SIDs allocated for TI-LFA backup paths.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: extend SRv6 TI-LFA topotests with backup path verification

Extend the isis_srv6_tilfa_topo1 topotest to verify the backup End.X
SID allocation and backup nexthop installation added in the preceding
commits for TI-LFA protection.

Add tests to verify backup SIDs are allocated with the backup flag set,
properly managed through topology changes, and cleaned up when no longer
needed. Add tests to verify backup nexthops are present in the RIB with
correct SRv6 segment stacks and that the backupIndex field references
valid backup nexthop indices.

Add tests to verify backup paths are pre-installed before failure and
that traffic switches to these paths when the primary link fails. Add
SID allocation stability verification that parses ISIS LFA debug output
to detect allocation loops or pending allocations.

Update kernel route verification to use iproute2 JSON output for reliable
parsing. Add helper functions to srv6_helper.py for backup SID queries,
backup nexthop verification, backup path capture and comparison, and
SID allocation stability monitoring.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
isisd, lib: add TI-LFA backup path support for primary End.X SIDs

When TI-LFA is enabled and backup End.X SIDs are computed for an
adjacency, the backup path information is now passed to zebra when
installing the corresponding primary End.X SID. This allows zebra
to pre-install backup nexthops that can be activated when the primary
link fails.

The seg6local_context structure is extended with fields to carry
backup path information including the backup nexthop address, backup
interface index, and a segment list for more complex repair paths.
The same fields are added to srv6_sid_ctx for consistency.

The zclient_send_localsid function is extended to check if backup
path information is present in the context. When available, it
configures the primary nexthop with the ZAPI_NEXTHOP_FLAG_HAS_BACKUP
flag and adds a backup nexthop entry with the alternate path details.

In isis_zebra_srv6_adj_sid_install, when installing a primary End.X
SID, the code now looks up whether a backup End.X SID exists for the
same adjacency. If found and it has computed backup nexthops from
TI-LFA, the first backup nexthop is extracted and used to populate
the backup path fields in the seg6local context before sending the
route to zebra.

This leverages zebra's existing backup nexthop infrastructure to
provide fast reroute capability for SRv6 End.X SIDs without requiring
kernel modifications.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
zebra: add seg6local fast reroute for SRv6 TI-LFA

Implement fast reroute for SRv6 TI-LFA seg6local routes in zebra.
When the primary interface goes down, zebra automatically switches
to the backup nexthop without waiting for ISIS to recalculate.

The implementation:
- Tracks seg6local routes with backup info in a hash table
- Links routes to interfaces via primary/backup ifindex
- On interface down: activates backup nexthop for affected routes
- On interface up: reverts routes to primary nexthop

This enables 10-50ms failover for seg6local routes with pre-computed
backup paths, compared to seconds when waiting for IGP reconvergence.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: add step 4 to test zebra fast reroute for SRv6 seg6local routes

Step 2 and step 4 both test failover behavior but exercise fundamentally
different code paths in the routing stack.

Step 2 uses vtysh to administratively shutdown an interface. This causes
ISIS to detect the adjacency loss through hello timeout, triggering a full
SPF recomputation and route recalculation. The entire ISIS reconvergence
process takes several seconds as the protocol withdraws LSPs, floods
updates to neighbors, and recomputes the shortest path tree.

Step 4 uses the kernel ip link command to bring the interface down. The
kernel immediately notifies zebra of the link state change through netlink.
Zebra then invokes the seg6local fast reroute mechanism which switches
affected routes to their pre-computed backup nexthops without waiting for
ISIS. This fast path completes in under 50 milliseconds because it bypasses
protocol reconvergence entirely and uses backup information that was
already installed in the forwarding plane.

The new tests verify that seg6local routes with backup nexthops are updated
immediately when the primary interface goes down, that traffic continues to
flow via the backup path, and that routes properly revert to the primary
nexthop when the interface recovers.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
zebra: add grace delay for seg6local fast reroute revert

When an interface goes down, zebra immediately switches seg6local routes
to their backup nexthops to minimize traffic loss. However, when the
interface comes back up, immediately reverting to the primary nexthop
can cause problems if the interface is unstable or flapping.

This change adds a configurable grace delay (default 1000ms) before
reverting seg6local routes from backup to primary nexthops. When the
interface comes up, zebra starts a timer instead of immediately
reverting. If the interface goes down again before the timer expires,
the timer is cancelled and routes remain on their backup paths. Only
when the timer expires does zebra revert routes to their primary
nexthops.

The grace delay provides several benefits. It prevents route flapping
when interfaces are unstable during link negotiation or when cables are
being reconnected. It gives the network time to stabilize before
switching back to the primary path. It also allows ISIS adjacencies to
re-establish before traffic is moved back.

The fast reroute to backup remains immediate (under 50ms) to minimize
packet loss during failures. Only the revert to primary is delayed to
ensure stability.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
zebra: add CLI command for seg6local fast reroute grace delay

Add a VTY command to configure the grace delay for seg6local fast
reroute under the segment-routing srv6 configuration node.

The command syntax is:
  fast-reroute grace-delay (0-60000)

Where the value is specified in milliseconds. Setting to 0 disables
the grace delay and reverts to primary immediately when the interface
comes up. The default value is 1000 milliseconds.

The no form resets to the default:
  no fast-reroute grace-delay

Example configuration:
  segment-routing
   srv6
    fast-reroute grace-delay 2000

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
doc: document SRv6 TI-LFA data plane fast reroute behavior

Add documentation explaining the two-level failover mechanism for SRv6
TI-LFA seg6local routes. The first level is protocol-based failover
where ISIS detects adjacency loss and reconverges, which takes several
seconds. The second level is data plane fast reroute where zebra
receives kernel interface down notifications and immediately switches
to backup nexthops in under 50 milliseconds.

The documentation describes how backup End.X SIDs are allocated and
passed from ISIS to zebra, how zebra tracks seg6local routes by their
primary interface, and how the grace delay timer prevents route
flapping when interfaces recover.

The new zebra CLI command for configuring the grace delay is documented
along with an example configuration. Examples showing how to verify
backup SID allocation using the show segment-routing srv6 command with
both text and JSON output formats are also provided.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

vjardin committed:
tests: enhance check_kernel_seg6_support to verify seg6local routes

Enhance check_kernel_seg6_support() to test if seg6local routes can
actually be installed, not just if the sysctl exists.

Signed-off-by: Vincent Jardin <vjardin@free.fr>

prvattem commented:
@vjardin - Are you looking into resolving the conflicts?

vjardin commented:
> @vjardin - Are you looking into resolving the conflicts?

I did suspend it for the time being due to some other on-going developments. 

What's your interest ?
